File: Java Programs/Chapter 2/Queue_LinkedList/src/queue_linkedlist/Queue_LinkedList.java

package queue_linkedlist;

//Program to implement Queue using linked list.
//Elements of any data type can be inserted into the queue.

import java.util.*;

//Node operations
class Node
{
	protected Object value;
	protected Node next;
	public Node()
	{
		value=null;
		next=null;
	}
	public Node(Object element)
	{
		value=element;
		next=null;
	}
	public Object getValue()
	{
		return value;
	}
	public Node getNext()
	{
		return next;
	}
	public void setNext(Node nxt)
	{
		next=nxt;
	}

}
//Queue operation
class Queue_link
{
	protected Node newNode;
	protected Node front;
	protected Node rear;
	
	public Queue_link()
	{
		front=null;
		rear=null;		
	}
	//Enqueue operation
	public void enqueue(Object element)
	{
		newNode=new Node(element);
		if(newNode==null)
		{
			System.out.println("Sufficient Memory not available to create node");
		}
		else
		{
			if(rear==null)
			{
				rear=newNode;
				front=newNode;
			}
			else
			{
				rear.setNext(newNode);
				rear=newNode;
			}
		}
		
	}
	//Dequeue operation
	public Object dequeue()
	{
		Object element;
		if(front==null)
		{
			System.out.println("Queue Underflow");
			element=null;
		}
		else
		{
			element=front.getValue();
			front=front.getNext();
			if(front==null)
			{
				rear=null;
			}
			
		}
		return element;
	}
	//Display queue
	public void displayQueue()
	{
		Node temp=front;
		System.out.print("Queue: ");
		while(temp!=null)
		{
			System.out.print(""+temp.getValue()+" | ");
			temp=temp.getNext();
		}
	}
	
}
public class Queue_LinkedList
{
	public static void main(String args[]) throws Exception
	{
		Scanner oin=new Scanner(System.in);
		Queue_link que=new Queue_link();
		Object element;
		int choice;
		
		while(true)
		{
			
			System.out.println("\n\nEnter the operation to be performed\n 1.Enqueue\n 2.Dequeue\n 3.Exit");
			System.out.print("Option: ");
			String s1=oin.nextLine();
			choice=Integer.parseInt(s1);
			if(choice==1)
			{
					System.out.print("Enter the element to be inserted: ");
					String s=oin.nextLine();
					element=(Object)s;
					que.enqueue(element);
					que.displayQueue();
			}
			else if(choice==2)
			{
				element=que.dequeue();
				if(element!=null)
				{
					System.out.println("Dequeued Element: "+element);
				}
				que.displayQueue();
			}
			else if(choice==3)
			{
				break;
			}
			else
			{
				System.out.println("Invalid Choice");
			}
				
		}
	}
	
}

File: Python Programs/unit2/QueueLinkedList.py
from __future__ import print_function

start = None
rear = None


class Node(object):

    def __init__(self, data=None, next_node=None):
        self.data = data
        self.next_node = next_node


def create():
    global start, rear
    print("Enter the value:")
    value = int(input())
    newNode = Node(data=value)
    if start is None:
        start = newNode
        rear = newNode
    else:
        rear.next_node = newNode
        rear = rear.next_node


def display():
    global start
    if start is None:
        print("Queue is empty:")
    else:
        temp = start
        while temp is not None:
            print(temp.data)
            temp = temp.next_node


def delete_node():
    global start
    num = start.data
    start = start.next_node
    return num


while True:
    print("1.INSERT\n2.DELETE\n3.EXIT")
    c = int(input("Enter your choice : "))
    if c == 1:
        create()
    elif c == 2:
        if start is None:
            print("Empty Queue linked list")
        else:
            print("Before deleting:")
            display()
            num = delete_node()
            print("after deleting:")
            display()
            print("Deleted %d" % num)

    elif c == 3:
        exit(0)
    else:
        print("Invalid choice")
